### PR TITLE
[`question_mark`]: also trigger on `return` statements

### PR DESCRIPTION
**Pull request by y21**
This fixes the false negative mentioned in #11993: the lint only used to check for `return` expressions, and not a statement containing a `return` expression (doesn't close the issue tho since there's still a useful suggestion that we could make, which is to suggest `.ok_or()?`/`.ok_or_else()?` for `else { return Err(..) }`)

changelog: [`question_mark`]: also trigger on `return` statements

### CODE REVIEW

**Comment by samueltardieu** [comment]
I like how it immediately flagged 3 instances in Clippy sources. I wonder how many will trigger in popular crates.

**Comment by samueltardieu** [comment]
From the top 500 crates, only 4 crates triggered this lint, totaling a number of 72 instances (most of them in the plotters-backend crate).

**Comment by y21** [comment]
Nice. [This comment](https://github.com/rust-lang/rust-clippy/issues/11993#issuecomment-1866196273) made me think that this would cause us to lint in a lot more cases due to rustfmt adding the semicolon and suppressing the lint that way, but double-checking now, it doesn't seem to be case that it does that.

Hopefully those 4 are legit warnings and not FPs? Also thanks for running lintcheck, I meant to do that myself but forgot 😅

**Comment by samueltardieu** [comment]
> Hopefully those 4 are legit warnings and not FPs?

Looks like it. I'm currently submitting a PR for the libc crate, and may submit others as well.

File: clippy_lints/src/question_mark.rs [hunk 1]
```diff
@@ -96,18 +96,32 @@ enum IfBlockType<'hir> {
...
         init: Some(init_expr),
         els: Some(els),
         ..
     }) = stmt.kind
-        && let Block {
-            stmts: &[],
-            expr: Some(els),
-            ..
-        } = els
+        && let Some(els) = find_let_else_ret_expression(els)
```
Comment (xFrednet) [inline clippy_lints/src/question_mark.rs:124]: This change looks good to me. Could you also add a check for comments, with [`span_contains_comment`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.span_contains_comment.html) and a test?

```rs
    let Some(x) = y else {
        // Roses are red,
        // violets are blue,
        // please keep this comment,
        // it's art, you know?
        return None;
    };
```

**Review by xFrednet — COMMENTED** [review]
LGTM, small NIT and then we can r+ this :D

File: clippy_lints/src/question_mark.rs (hunk 1 above)
Comment (y21) [reply clippy_lints/src/question_mark.rs]: Good idea! Fixed

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Perfection! And what a masterpiece.

@bors r+

**Pull request merged**